### PR TITLE
Add backup-version file in backup tarball

### LINKED ISSUE #134: Backup tarball should include backup version in a file
Let's put it in `backup-version`

### PR DESCRIPTION
**Pull request by wwitzel3**
Closes #134

### CODE REVIEW

File: pkg/backup/backup.go [hunk 1]
```diff
@@ -292,6 +297,23 @@ func (kb *kubernetesBackupper) Backup(logger logrus.FieldLogger, backupRequest *
 	return err
 }
 
+func (kb *kubernetesBackupper) writeBackupVersion(tw *tar.Writer) error {
+	hdr := &tar.Header{
+		Name:     "backup-version",
+		Size:     int64(len(buildinfo.Version)),
```
Comment (ncdc) [inline pkg/backup/backup.go]: This isn't the ark version. It needs to be a "backup version" that starts at something like 1 or 1.0.0 and then we'll manually increment as needed. And eventually have code that handles restoring different backup versions.

Comment (wwitzel3) [reply pkg/backup/backup.go]: Ahh, ok. So we'll just store that in a const then?

File: pkg/backup/backup.go [hunk 2]
```diff
@@ -41,6 +41,9 @@ import (
 	kubeutil "github.com/heptio/ark/pkg/util/kube"
 )
 
+// BackupVersion is the current backup version for Ark.
+const BackupVersion = "1.0.0"
```
Comment (ncdc) [inline pkg/backup/backup.go]: Should probably be `1`

Comment (ncdc) [reply pkg/backup/backup.go]: https://github.com/heptio/ark/blob/c34e3b91276f7a2787264deb5f6be345990c7c8d/pkg/controller/backup_controller.go#L55

**Comment by wwitzel3** [comment]
@ncdc PTAL

File: pkg/backup/backup.go [hunk 3]
```diff
@@ -292,6 +300,25 @@ func (kb *kubernetesBackupper) Backup(logger logrus.FieldLogger, backupRequest *
...
+		Name:     "backup-version",
+		Size:     int64(len(versionString)),
+		Typeflag: tar.TypeReg,
+		Mode:     0644,
+		ModTime:  time.Now(),
+	}
+	if err := tw.WriteHeader(hdr); err != nil {
+		return err
+	}
+	if _, err := tw.Write([]byte(versionString)); err != nil {
```
Comment (ncdc) [inline pkg/backup/backup.go]: Do you want to include a newline? `fmt.Fprintln(tw, versionString)`, or even `fmt.Fprintf(tw, "%d\n", BackupVersion)`?

**Comment by skriss** [comment]
Now that I'm thinking about this PR - we will now store some metadata as objects directly in object storage (e.g. `ark-backup.json`), and some (e.g. `backup-version`) as files within the tarball. Do we have a principled stance on what should go where? Would it make sense to duplicate any of it?

**Comment by nrb** [comment]
@skriss That's a good question on the duplication. My current thought is that the backup version should be in the tarball, so that it could be processed if moved around. Information that's useful for periodic checks should be outside the tarball, though, since we don't want to download and extract them all on a regular basis.

**Comment by wwitzel3** [comment]
I remember during the original discussion we wanted to intentionally keep `backup-version` in the tarball for the reason mentioned by @nrb

**Comment by ncdc** [comment]
Nolan is correct - we want the `backup-version` somewhere in the tarball so that we can identify its format version even if we don't have the json to go with it. Maybe we should also include the json in it as well...

👍 to `metadata/backup-version` or even just `metadata/version`

**Comment by skriss** [comment]
SGTM for now, though we probably need more thinking/work about how a tarball without an `ark-backup.json` file would be portable, if that's a goal.  Can address later.

**Comment by ncdc** [comment]
I guess it probably can't be 100% portable, since we have other information in the bucket (snapshot details, logs)

**Comment by skriss** [comment]
this lgtm pending moving the file into a `metadata/` dir

**Comment by wwitzel3** [comment]
Ok, I think this is ready to go now. PTAL @skriss @ncdc

**Comment by wwitzel3** [comment]
This is ready to merge if everyone is happy with it.

File: pkg/backup/backup.go [hunk 4]
```diff
@@ -292,6 +300,26 @@ func (kb *kubernetesBackupper) Backup(logger logrus.FieldLogger, backupRequest *
...
+
+	hdr := &tar.Header{
+		Name:     versionFile,
+		Size:     int64(len(versionString)),
+		Typeflag: tar.TypeReg,
+		Mode:     0644,
+		ModTime:  time.Now(),
+	}
+	if err := tw.WriteHeader(hdr); err != nil {
+		return err
```
Comment (ncdc) [inline pkg/backup/backup.go]: I think we want `errors.Wrap` here

File: pkg/backup/backup.go [hunk 5]
```diff
@@ -292,6 +300,26 @@ func (kb *kubernetesBackupper) Backup(logger logrus.FieldLogger, backupRequest *
...
+	}
+	if _, err := tw.Write([]byte(versionString)); err != nil {
+		return err
```
Comment (ncdc) [inline pkg/backup/backup.go]: errors.Wrap

**Comment by skriss** [comment]
@ncdc I think we were just waiting for your approval here since you had a comment - merge if happy :)

**Comment by ncdc** [comment]
Didn't realize y'all were waiting on me!

**Pull request merged**